core/search/service.py

```diff
@@ -0,0 +1,330 @@
+        """Atomically set the read pointer to the empty index for the base name."""
+
+    @abstractmethod
+    def index_create(self, base_name: str, revision: SearchSchemaRevision) -> None:
+        """Atomically create an index for the given base name and revision."""
+
+    @abstractmethod
+    def indexes_created(self) -> List[str]:
+        """A log of all the indexes that have been created by this client service."""
+
+    @abstractmethod
+    def index_is_populated(
+        self, base_name: str, revision: SearchSchemaRevision
+    ) -> bool:
+        """Return True if the index for the given base name and revision has been populated."""
+
+    @abstractmethod
+    def index_set_mapping(self, base_name: str, revision: SearchSchemaRevision) -> None:
+        """Set the schema mappings for the given index."""
+
+    @abstractmethod
+    def index_submit_documents(
+        self,
+        pointer: str,
+        documents: Iterable[dict],
+    ) -> List[SearchServiceFailedDocument]:
+        """Submit search documents to the given index."""
+
+    @abstractmethod
+    def write_pointer_set(self, base_name: str, revision: SearchSchemaRevision) -> None:
+        """Atomically set the write pointer to the index for the given revision and base name."""
+
+    @abstractmethod
+    def refresh(self):
+        """Synchronously refresh the service and wait for changes to be completed."""
+
+    @abstractmethod
+    def index_clear_documents(self, pointer: str):
+        """Clear all search documents in the given index."""
+
+    @abstractmethod
+    def search_client(self) -> Search:
+        """Return the underlying search client."""
+
+    @abstractmethod
+    def search_multi_client(self) -> MultiSearch:
+        """Return the underlying search client."""
+
+    @abstractmethod
+    def index_remove_document(self, pointer: str, id: int):
+        """Remove a specific document from the given index."""
+
+
+class SearchServiceOpensearch1(SearchService):
+    """The real Opensearch 1.x service."""
+
+    def __init__(self, client: OpenSearch):
+        self._logger = logging.getLogger(SearchServiceOpensearch1.__name__)
+        self._client = client
+        self._search = Search(using=self._client)
+        self._multi_search = MultiSearch(using=self._client)
+        self._indexes_created: List[str] = []
+
+        # Documents are not allowed to automatically create indexes.
+        self._client.cluster.put_settings(
+            body={"persistent": {"action": {"auto_create_index": "false"}}}
+        )
+
+    def indexes_created(self) -> List[str]:
+        return self._indexes_created
+
+    def write_pointer(self, base_name: str) -> Optional[SearchWritePointer]:
+        try:
+            result: dict = self._client.indices.get_alias(
+                name=self.write_pointer_name(base_name)
+            )
+            for name in result.keys():
+                match = re.search(f"{base_name}-v([0-9]+)", string=name)
```

Comment:
I've been so buried in test failures that I've forgotten why it was like this. I'm going to adjust these patterns today as I think they need to be more explicit. Our read/write pointers should only ever point to the versioned indices, or the empty index. They're not user-serviceable parts and should never point to anything else.

Reply:
Actually, this is correct:

The asymmetry is because it's always safe to _read_ from an index (even the empty index), but it's not safe to _write_ to an index (as in, submit documents and/or update the mappings) unless you know the structure of the mappings. Hence, the `write_pointer` method either returns the versioned index name, or it returns nothing at all so that you can't write to an index.

Reply:
Cool. That's some implicit behaviour, maybe we can add some comments explaining it out in the method.

Reply:
Will do!